src/kernels/MIOpenDiag.cpp

```diff
@@ -0,0 +1,62 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename TIO>
+__device__ void Diag2dForwardKernel(const TIO* input,
+                                    TIO* output,
+                                    long N,
+                                    long offset,
+                                    tensor_view_t<2> input_tv,
+                                    tensor_view_t<1> output_tv)
+{
+    size_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+    if(gid >= N)
+        return;
+
+    long input_stride_0 = input_tv.stride[0];
+    long input_stride_1 = input_tv.stride[1];
+
+    long input_idx     = gid * (input_stride_0 + input_stride_1) + offset;
+    long output_idx    = gid * output_tv.stride[0];
+    output[output_idx] = input[input_idx];
```

Comment:
Looking at moDNN's kernel code, input_idx is calculated and then the tensor is accessed with R_GET_2D_VAL. And R_GET_2D_VAL calculates the index once again with IDX_TO_TV2D_IDX.
If the tensor is contiguous, you can access the tensor directly with idx, but if it is non-contiguous, you must access it with size and stride.

Reply:
I have updated the kernel implementation